### DISCUSSION TITLE
mterm that accepts _u as an argument

### DISCUSSION THREAD

**Discussion by pavaninguva** [do-mpc/do-mpc#231] [Q&A] [answered]
Hi, I was implementing a simple CSTR-like model for the system and to set the objective function such that I maximise productivity, my `lterm` looks like:
```
lterm = -model.u["F"]*(model.x["C_s"]+model.x["C_d"]),
```
where F is the flowrate in and out of the reactor and is an input `_u` and C_s and C_d are my products concentrations and are states `_x` of the system. From the documentation, I see that I am unable to implement the same expression for the `mterm` since it only accepts `_x` and `_p` terms. 

This maybe a naive question, but is it possible to have an `mterm` that accepts `_u` and if not, would simply having an `lterm` with `mterm = DM(0)` a valid formulation for the problem. I am not sure if the lack of an `mterm` might create any issues. 

Thanks for the help!

**Comment by 4flixt** [comment, marked as answer]
Dear @pavaninguva,

I start with your second question: Yes, it is possible to have a blank lterm or mterm by choosing the expression ``mterm = DM(0)``. I meant to make this even easier in the future #86 addresses the issue. There might be cases were a blank mterm is useful. 

You first question: Due to the nature of the optimal control problem it can never be possible to have an input in the LAST time-step of the prediction horizon. Imagine a simple example with a horizon of N=1. You have ``u_0`` and ``x_0`` which allows to compute the state ``x_1``. You now have an ``mterm(x_0, u_0)`` and an ``lterm(x_1)``. If you would have another input available (``u_1``) you could (and should) also compute ``x_2``. You would therefore have a horizon of N=2 and the same problem as before.

In some way having ``lterm=DM(0)`` is therefore exactly what you want. You simply ignore the last state at ``x_N+1`` for your objective.

P.S.: If you are not sure if you are reporting a bug (or simply have a question) please **don't raise an issue** but post a question to the [discussions](https://github.com/do-mpc/do-mpc/discussions) tab.

Best,
4flixt

**Comment by pavaninguva** [comment]
Hi @4flixt Thank you so much for your kind response and apologies for the delay. Only recently revisited my work here. 

The fix you gave works fine. The case I am exploring using an economic cost function where the Lagrange term is the product of `_u` and `_x`. I would imagine this sort of case would be quite relevant for many process / chemical systems. 

Regarding the 2nd question I raised, thank you for the clarification. After becoming more familiar with the MPC literature, I have a better understanding of the purpose of the Meyer term and why it doesnt make sense for it be a function of `_u`.
